### PR TITLE
[GEOT-7429] Vector Mosaic add native retyping support when query is exclusively granule or index fields

### PR DESCRIPTION
**Pull request by turingtestfail**
[GEOT-7429](https://osgeo-org.atlassian.net/browse/GEOT-7429) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

…xclusively granule or index fields

clean

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [x] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

**Review by aaime — CHANGES_REQUESTED** [review]

File: modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureSource.java [hunk 1]
```diff
@@ -414,7 +451,7 @@ protected void populateGranuleTypeName(VectorMosaicGranule granule, DataStore da
     }
 
     @Override
-    protected boolean canFilter() {
+    protected boolean canRetype() {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureSource.java]: Uggh, merge issue.  Fixed.

**Review by aaime — APPROVED** [review]

**Comment by aaime** [comment]
Looking at the code, it seems that the delegate readers are still reading more attributes than strictly necessary. Like, if only one attribute is needed from the granule store, then only that should be read (setting up the query for it accordingly). The delegate case is interesting, as you'll still need at least the location/property attribute to open the delegates, forcing a bit of re-typing after the fact.

**Comment by turingtestfail** [comment]
> Looking at the code, it seems that the delegate readers are still reading more attributes than strictly necessary. Like, if only one attribute is needed from the granule store, then only that should be read (setting up the query for it accordingly). The delegate case is interesting, as you'll still need at least the location/property attribute to open the delegates, forcing a bit of re-typing after the fact.

Yes - redone to limit the attributes for both the delegate and granule queries.

**Comment by aaime** [comment]
Needs to be rebased and conflicts resolved.

**Comment by turingtestfail** [comment]
> Needs to be rebased and conflicts resolved.

Done

**Comment by aaime** [comment]
The backport to `28.x` failed:
```
The process '/usr/bin/git' failed with exit code 1
```
<details>
<summary>stderr</summary>

```
error: could not apply 533bfe3f48... [GEOT-7429] Vector Mosaic add native retyping support when query is exclusively granule or index fields
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git cherry-pick --continue".
hint: You can instead skip this commit with "git cherry-pick --skip".
hint: To abort and get back to the state before "git cherry-pick",
hint: run "git cherry-pick --abort".

```
</details>

<details>
<summary>stdout</summary>

```
Auto-merging modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureReader.java
CONFLICT (content): Merge conflict in modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureReader.java
Auto-merging modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureSource.java
Auto-merging modules/unsupported/vector-mosaic/src/test/java/org/geotools/vectormosaic/VectorMosaicFeatureSourceTest.java
CONFLICT (content): Merge conflict in modules/unsupported/vector-mosaic/src/test/java/org/geotools/vectormosaic/VectorMosaicFeatureSourceTest.java

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-28.x 28.x
# Navigate to the new working tree
cd .worktrees/backport-28.x
# Create a new branch
git switch --create backport-4459-to-28.x
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick 533bfe3f48250c802e2e4ab2d6969fa9bb4a8640,92b097e95be6076e8bfafd6214f1f1276c0060e7
# Push it to GitHub
git push --set-upstream origin backport-4459-to-28.x
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-28.x
```
Then, create a pull request where the `base` branch is `28.x` and the `compare`/`head` branch is `backport-4459-to-28.x`.

**Comment by aaime** [comment]
The backport to `29.x` failed:
```
The process '/usr/bin/git' failed with exit code 1
```
<details>
<summary>stderr</summary>

```
error: could not apply 533bfe3f48... [GEOT-7429] Vector Mosaic add native retyping support when query is exclusively granule or index fields
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git cherry-pick --continue".
hint: You can instead skip this commit with "git cherry-pick --skip".
hint: To abort and get back to the state before "git cherry-pick",
hint: run "git cherry-pick --abort".

```
</details>

<details>
<summary>stdout</summary>

```
Auto-merging modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureReader.java
CONFLICT (content): Merge conflict in modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureReader.java
Auto-merging modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureSource.java
Auto-merging modules/unsupported/vector-mosaic/src/test/java/org/geotools/vectormosaic/VectorMosaicFeatureSourceTest.java
CONFLICT (content): Merge conflict in modules/unsupported/vector-mosaic/src/test/java/org/geotools/vectormosaic/VectorMosaicFeatureSourceTest.java

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-29.x 29.x
# Navigate to the new working tree
cd .worktrees/backport-29.x
# Create a new branch
git switch --create backport-4459-to-29.x
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick 533bfe3f48250c802e2e4ab2d6969fa9bb4a8640,92b097e95be6076e8bfafd6214f1f1276c0060e7
# Push it to GitHub
git push --set-upstream origin backport-4459-to-29.x
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-29.x
```
Then, create a pull request where the `base` branch is `29.x` and the `compare`/`head` branch is `backport-4459-to-29.x`.

**Pull request merged**